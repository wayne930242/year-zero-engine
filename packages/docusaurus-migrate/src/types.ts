/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export type RawData = {
  header?: string;
  content?: string;
};

export type Data = {
  metadata: {[key: string]: string};
  rawContent: string;
};

export type ClassicPresetEntries = {
  docs: {[key: string]: any};
  blog: {[key: string]: any};
  theme: {[key: string]: any};
};

export type SidebarEntries = {
  [key: string]:
    | Record<string, unknown>
    | Array<Record<string, unknown> | string>;
};

export interface VersionTwoConfig {
  baseUrl: string;
  favicon: string;
  tagline?: string;
  title: string;
  url: string;
  organizationName?: string;
  projectName?: string;
  noIndex?: boolean;
  githubHost?: string;
  onBrokenLinks: string;
  onBrokenMarkdownLinks: string;
  plugins: Array<[string, {[key: string]: any}]>;
  themes?: [];
  presets: [[string, ClassicPresetEntries]];
  themeConfig: {
    gtag?: {
      trackingID?: string;
    };
    navbar: {
      title?: string;
      logo?: {
        src?: string;
      };
      items: Array<Record<string, unknown> | null>;
    };
    image?: string;
    footer: {
      links: Array<{
        title: string;
        items: Array<{
          label: string;
          to: string;
        }>;
      }>;
      copyright?: string;
      logo: {
        src?: string;
      };
    };
    algolia?: Record<string, unknown>;
  };
  customFields: {
    [key: string]: unknown;
  };
  scripts?: (
    | string
    | {
        src: string;
        [key: string]: unknown;
      }
  )[];
  stylesheets?: (
    | string
    | {
        href: string;
        [key: string]: unknown;
      }
  )[];
}

export type VersionOneConfig = {
  title?: string;
  tagline?: string;
  url?: string;
  baseUrl?: string;
  defaultVersionShown?: string;
  organizationName?: string;
  projectName?: string;
  noIndex?: boolean;
  headerLinks?: Array<any>;
  headerIcon?: string;
  favicon?: string;
  colors?: any;
  copyright?: string;
  editUrl?: string;
  customDocsPath?: string;
  users?: Array<Record<string, unknown>>;
  disableHeaderTitle?: string;
  disableTitleTagline?: string;
  separateCss?: Array<Record<string, unknown>>;
  footerIcon?: string;
  translationRecruitingLink?: string;
  algolia?: Record<string, unknown>;
  gaTrackingId?: string;
  highlight?: Record<string, unknown>;
  markdownPlugins?: Array<() => void>;
  scripts?: Array<{src: string; [key: string]: any} | string>;
  stylesheets?: Array<{href: string; [key: string]: any} | string>;
  facebookAppId?: string;
  facebookComments?: true;
  facebookPixelId?: string;
  twitter?: string;
  twitterUsername?: string;
  twitterImage?: string;
  ogImage?: string;
  cleanUrl?: boolean;
  scrollToTop?: boolean;
  scrollToTopOptions?: Record<string, unknown>;
};
