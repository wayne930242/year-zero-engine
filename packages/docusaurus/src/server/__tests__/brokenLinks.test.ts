/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {
  getBrokenLinksErrorMessage,
  getAllBrokenLinks,
  filterExistingFileLinks,
} from '../brokenLinks';
import {RouteConfig} from '@docusaurus/types';

describe('brokenLinks', () => {
  test('getBrokenLinksErrorMessage', async () => {
    const message = getBrokenLinksErrorMessage({
      '/docs/mySourcePage': [
        {link: './myBrokenLink', resolvedLink: '/docs/myBrokenLink'},
        {link: '../otherBrokenLink', resolvedLink: '/otherBrokenLink'},
      ],
      '/otherSourcePage': [{link: '/badLink', resolvedLink: '/badLink'}],
    });
    expect(message).toMatchSnapshot();
  });

  test('getBrokenLinksErrorMessage with potential layout broken links', async () => {
    const frequentLink = [
      {
        link: './myBrokenLinkFrequent1',
        resolvedLink: '/docs/myBrokenLinkFrequent1',
      },
      {
        link: './myBrokenLinkFrequent2',
        resolvedLink: '/docs/myBrokenLinkFrequent2',
      },
    ];
    const infrequentLink = [
      {
        link: './myBrokenLinkInfrequent1',
        resolvedLink: '/docs/myBrokenLinkInfrequent1',
      },
      {
        link: './myBrokenLinkInfrequent2',
        resolvedLink: '/docs/myBrokenLinkInfrequent2',
      },
    ];

    const message = getBrokenLinksErrorMessage({
      '/docs/page1': [...frequentLink],
      '/docs/page2': [...frequentLink, ...infrequentLink],
      '/docs/page3': [...frequentLink],
      '/docs/page4': [...frequentLink, ...infrequentLink],
      '/docs/page5': [...frequentLink],
      '/docs/page6': [...frequentLink],
    });
    expect(message).toMatchSnapshot();
  });

  test('getAllBrokenLinks', async () => {
    const routes: RouteConfig[] = [
      {
        path: '/docs',
        component: '',
        routes: [
          {path: '/docs/someDoc', component: ''},
          {path: '/docs/someOtherDoc', component: ''},
        ],
      },
      {
        path: '/community',
        component: '',
      },
      {
        path: '*',
        component: '',
      },
    ];

    const allCollectedLinks = {
      '/docs/someDoc': [
        // Good links
        './someOtherDoc#someHash',
        '/docs/someOtherDoc?someQueryString=true#someHash',
        '../docs/someOtherDoc?someQueryString=true',
        '../docs/someOtherDoc#someHash',
        // Bad links
        '../someOtherDoc',
        './docThatDoesNotExist',
        './badRelativeLink',
        '../badRelativeLink',
      ],
      '/community': [
        // Good links
        '/docs/someDoc',
        '/docs/someOtherDoc#someHash',
        './docs/someDoc#someHash',
        './docs/someOtherDoc',
        // Bad links
        '/someOtherDoc',
        '/badLink',
        './badLink',
      ],
    };

    const expectedBrokenLinks = {
      '/docs/someDoc': [
        {
          link: '../someOtherDoc',
          resolvedLink: '/someOtherDoc',
        },
        {
          link: './docThatDoesNotExist',
          resolvedLink: '/docs/docThatDoesNotExist',
        },
        {
          link: './badRelativeLink',
          resolvedLink: '/docs/badRelativeLink',
        },
        {
          link: '../badRelativeLink',
          resolvedLink: '/badRelativeLink',
        },
      ],
      '/community': [
        {
          link: '/someOtherDoc',
          resolvedLink: '/someOtherDoc',
        },
        {
          link: '/badLink',
          resolvedLink: '/badLink',
        },
        {
          link: './badLink',
          resolvedLink: '/badLink',
        },
      ],
    };

    expect(getAllBrokenLinks({allCollectedLinks, routes})).toEqual(
      expectedBrokenLinks,
    );
  });

  test('filterExistingFileLinks', async () => {
    const link1 = '/link1';
    const link2 = '/docs/link2';
    const link3 = '/hey/link3';

    const linkToJavadoc1 = '/javadoc';
    const linkToJavadoc2 = '/javadoc/';
    const linkToJavadoc3 = '/javadoc/index.html';

    const linkToZipFile = '/files/file.zip';
    const linkToHtmlFile1 = '/files/hey.html';
    const linkToHtmlFile2 = '/files/hey';

    const linkToEmptyFolder1 = '/emptyFolder';
    const linkToEmptyFolder2 = '/emptyFolder/';

    const allCollectedLinks = {
      '/page1': [
        link1,
        linkToHtmlFile1,
        linkToJavadoc1,
        linkToHtmlFile2,
        linkToJavadoc3,
        linkToEmptyFolder1,
      ],
      '/page2': [
        link2,
        linkToEmptyFolder2,
        linkToJavadoc2,
        link3,
        linkToJavadoc3,
        linkToZipFile,
      ],
    };

    const allCollectedLinksFiltered = {
      '/page1': [
        link1,
        // linkToHtmlFile1,
        // linkToJavadoc1,
        // linkToHtmlFile2,
        // linkToJavadoc3,
        linkToEmptyFolder1, // not filtered !
      ],
      '/page2': [
        link2,
        linkToEmptyFolder2, // not filtered !
        // linkToJavadoc2,
        link3,
        // linkToJavadoc3,
        // linkToZipFile,
      ],
    };

    const result = await filterExistingFileLinks({
      baseUrl: '/',
      outDir: path.resolve(__dirname, '__fixtures__/brokenLinks/outDir'),
      allCollectedLinks,
    });
    expect(result).toEqual(allCollectedLinksFiltered);
  });
});
