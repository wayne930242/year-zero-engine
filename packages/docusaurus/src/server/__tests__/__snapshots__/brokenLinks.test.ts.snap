// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`brokenLinks getBrokenLinksErrorMessage 1`] = `
"Docusaurus found broken links!

Please check the pages of your site in the list bellow, and make sure you don't reference any path that does not exist.
Note: it's possible to ignore broken links with the 'onBrokenLinks' Docusaurus configuration, and let the build pass.

Exhaustive list of all broken links found:

- On source page path = /docs/mySourcePage:
   -> linking to ./myBrokenLink (resolved as: /docs/myBrokenLink)
   -> linking to ../otherBrokenLink (resolved as: /otherBrokenLink)

- On source page path = /otherSourcePage:
   -> linking to /badLink
"
`;

exports[`brokenLinks getBrokenLinksErrorMessage with potential layout broken links 1`] = `
"Docusaurus found broken links!

Please check the pages of your site in the list bellow, and make sure you don't reference any path that does not exist.
Note: it's possible to ignore broken links with the 'onBrokenLinks' Docusaurus configuration, and let the build pass.

It looks like some of the broken links we found appear in many pages of your site.
Maybe those broken links appear on all pages through your site layout?
We recommend that you check your theme configuration for such links (particularly, theme navbar and footer).
Frequent broken links are linking to: 
- ./myBrokenLinkFrequent1
- ./myBrokenLinkFrequent2


Exhaustive list of all broken links found:

- On source page path = /docs/page1:
   -> linking to ./myBrokenLinkFrequent1 (resolved as: /docs/myBrokenLinkFrequent1)
   -> linking to ./myBrokenLinkFrequent2 (resolved as: /docs/myBrokenLinkFrequent2)

- On source page path = /docs/page2:
   -> linking to ./myBrokenLinkFrequent1 (resolved as: /docs/myBrokenLinkFrequent1)
   -> linking to ./myBrokenLinkFrequent2 (resolved as: /docs/myBrokenLinkFrequent2)
   -> linking to ./myBrokenLinkInfrequent1 (resolved as: /docs/myBrokenLinkInfrequent1)
   -> linking to ./myBrokenLinkInfrequent2 (resolved as: /docs/myBrokenLinkInfrequent2)

- On source page path = /docs/page3:
   -> linking to ./myBrokenLinkFrequent1 (resolved as: /docs/myBrokenLinkFrequent1)
   -> linking to ./myBrokenLinkFrequent2 (resolved as: /docs/myBrokenLinkFrequent2)

- On source page path = /docs/page4:
   -> linking to ./myBrokenLinkFrequent1 (resolved as: /docs/myBrokenLinkFrequent1)
   -> linking to ./myBrokenLinkFrequent2 (resolved as: /docs/myBrokenLinkFrequent2)
   -> linking to ./myBrokenLinkInfrequent1 (resolved as: /docs/myBrokenLinkInfrequent1)
   -> linking to ./myBrokenLinkInfrequent2 (resolved as: /docs/myBrokenLinkInfrequent2)

- On source page path = /docs/page5:
   -> linking to ./myBrokenLinkFrequent1 (resolved as: /docs/myBrokenLinkFrequent1)
   -> linking to ./myBrokenLinkFrequent2 (resolved as: /docs/myBrokenLinkFrequent2)

- On source page path = /docs/page6:
   -> linking to ./myBrokenLinkFrequent1 (resolved as: /docs/myBrokenLinkFrequent1)
   -> linking to ./myBrokenLinkFrequent2 (resolved as: /docs/myBrokenLinkFrequent2)
"
`;
